File: src/main/java/pl/qprogramming/daily/service/weather/WeatherService.java
package pl.qprogramming.daily.service.weather;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;
import pl.qprogramming.daily.dto.AirQualityData;
import pl.qprogramming.daily.dto.GeocodingResult;
import pl.qprogramming.daily.dto.WeatherData;
import pl.qprogramming.daily.dto.WeatherForecast;
import pl.qprogramming.daily.service.weather.mapper.WeatherMapper;
import pl.qprogramming.daily.service.weather.model.GeocodingResponse;
import pl.qprogramming.daily.service.weather.model.OpenMeteoAirQuality;
import pl.qprogramming.daily.service.weather.model.OpenMeteoWeatherResponse;

import static pl.qprogramming.daily.service.weather.WeatherConstants.*;

/**
 * Service for interacting with Open-Meteo weather APIs.
 * <p>
 * This service provides methods to access various weather-related data from Open-Meteo APIs:
 * <ul>
 *     <li>Geocoding - convert city names to geographic coordinates</li>
 *     <li>Current weather - get current weather conditions for a location</li>
 *     <li>Weather forecast - get daily and hourly weather forecasts</li>
 *     <li>Air quality - get air quality metrics for a location</li>
 * </ul>
 * All methods use caching to minimize API calls and improve performance.
 * </p>
 */
@Service
@Slf4j
public class WeatherService {

    private final RestTemplate restTemplate;
    private final WeatherMapper weatherMapper;

    /**
     * Constructor for WeatherService.
     * Initializes a RestTemplate for making HTTP requests to weather APIs.
     *
     * @param weatherMapper Mapper for converting between API response models and DTOs
     */
    public WeatherService(WeatherMapper weatherMapper) {
        this.restTemplate = new RestTemplate();
        this.weatherMapper = weatherMapper;
    }

    /**
     * Gets geocoding data for a city name.
     *
     * @param cityName Name of the city to geocode
     * @param language Language code for the response (defaults to "en" if null)
     * @param count Maximum number of results to return
     * @return GeocodingResult containing location data or null if not found or error occurs
     */
    @Cacheable(value = GEOCODING_CACHE, key = "#cityName + '-' + #language")
    public GeocodingResult geocodeLocation(String cityName, String language, int count) {
        if (cityName == null || cityName.trim().isEmpty()) {
            return null;
        }

        // Build the URL manually to preserve Polish characters instead of encoding them
        String baseUrl = OPEN_METEO_GEOCODING_URL + "?";
        String params = PARAM_NAME + "=" + cityName + "&" +
                PARAM_LANGUAGE + "=" + (language != null ? language : DEFAULT_LANGUAGE) + "&" +
                PARAM_COUNT + "=" + count;
        String url = baseUrl + params;

        try {
            val response = restTemplate.getForObject(url, GeocodingResponse.class);
            log.debug("Response from Open-Meteo Geocoding API: {}", response);
            if (response != null && response.getResults() != null && !response.getResults().isEmpty()) {
                val result = response.getResults().get(0);
                return weatherMapper.toGeocodingResponse(result);
            }
            return null;
        } catch (Exception e) {
            log.error("Error geocoding location '{}': {}", cityName, e.getMessage());
            return null;
        }
    }

    /**
     * Gets current weather data for a geographic location.
     * <p>
     * This method retrieves the current weather conditions for the specified
     * coordinates by calling the Open-Meteo Forecast API. The data includes
     * temperature, weather code, wind speed, and humidity. Results are cached
     * to reduce API calls.
     * </p>
     *
     * @param latitude Latitude of the location
     * @param longitude Longitude of the location
     * @return WeatherData containing current weather conditions or null if error occurs
     */
    @Cacheable(value = CURRENT_WEATHER_CACHE, key = "#latitude + '-' + #longitude")
    public WeatherData getCurrentWeather(double latitude, double longitude) {
        try {
            String url = UriComponentsBuilder.fromUriString(OPEN_METEO_FORECAST_URL)
                    .queryParam(PARAM_LATITUDE, latitude)
                    .queryParam(PARAM_LONGITUDE, longitude)
                    .queryParam(PARAM_CURRENT, CURRENT_WEATHER_PARAMS)
                    .queryParam(PARAM_TIMEZONE, TIMEZONE_AUTO)
                    .encode()
                    .toUriString();

            val response = restTemplate.getForObject(url, OpenMeteoWeatherResponse.class);
            log.debug("Response from Open-Meteo Current Weather API: {}", response);
            if (response == null || response.getCurrent() == null) {
                return null;
            }
            return weatherMapper.toWeatherData(response);
        } catch (Exception e) {
            log.error("Error fetching current weather for lat: {}, lon: {}: {}", latitude, longitude, e.getMessage());
            return null;
        }
    }

    /**
     * Gets weather forecast for a geographic location.
     * <p>
     * This method retrieves both daily and hourly weather forecasts for the
     * specified coordinates by calling the Open-Meteo Forecast API. The data
     * includes temperature, weather code, wind speed, and humidity. The forecast
     * can be limited by number of days and hours. Results are cached to reduce API calls.
     * </p>
     *
     * @param latitude Latitude of the location
     * @param longitude Longitude of the location
     * @param days Number of days to forecast
     * @param hours Number of hours to forecast
     * @return WeatherForecast containing daily and hourly forecast data or null if error occurs
     */
    @Cacheable(value = FORECAST_CACHE, key = "#latitude + '-' + #longitude + '-' + #days")
    public WeatherForecast getWeatherForecast(double latitude, double longitude, Integer days, Integer hours) {
        try {
            String url = UriComponentsBuilder.fromUriString(OPEN_METEO_FORECAST_URL)
                    .queryParam(PARAM_LATITUDE, latitude)
                    .queryParam(PARAM_LONGITUDE, longitude)
                    .queryParam(PARAM_CURRENT, CURRENT_WEATHER_PARAMS)
                    .queryParam(PARAM_DAILY, DAILY_WEATHER_PARAMS)
                    .queryParam(PARAM_HOURLY, HOURLY_WEATHER_PARAMS)
                    .queryParam(PARAM_FORECAST_DAYS, days)
                    .queryParam(PARAM_FORECAST_HOURS, hours)
                    .queryParam(PARAM_TIMEZONE, TIMEZONE_AUTO)
                    .encode()
                    .toUriString();

            val response = restTemplate.getForObject(url, OpenMeteoWeatherResponse.class);
            log.debug("Response from Open-Meteo Forecast: {}", response);
            if (response == null || response.getDaily() == null) {
                return null;
            }
            // Create context wrapper objects and pass them to the mapper
            val daysContext = new WeatherMapper.DaysContext(days);
            val hoursContext = new WeatherMapper.HoursContext(hours);
            return weatherMapper.toWeatherForecast(response, daysContext, hoursContext);
        } catch (Exception e) {
            log.error("Error fetching weather forecast for lat: {}, lon: {}, days: {}: {}", latitude, longitude, days, e.getMessage());
            return null;
        }
    }

    /**
     * Gets air quality data for a geographic location.
     * <p>
     * This method retrieves current air quality metrics for the specified
     * coordinates by calling the Open-Meteo Air Quality API. The data includes
     * PM10, PM2.5, and European Air Quality Index. Results are cached to reduce API calls.
     * </p>
     *
     * @param latitude Latitude of the location
     * @param longitude Longitude of the location
     * @return AirQualityData containing air quality metrics or null if error occurs
     */
    @Cacheable(value = AIR_QUALITY_CACHE, key = "#latitude + '-' + #longitude")
    public AirQualityData getAirQuality(double latitude, double longitude) {
        try {
            String url = UriComponentsBuilder.fromUriString(OPEN_METEO_AIR_QUALITY_URL)
                    .queryParam(PARAM_LATITUDE, latitude)
                    .queryParam(PARAM_LONGITUDE, longitude)
                    .queryParam(PARAM_CURRENT, AIR_QUALITY_PARAMS)
                    .encode()
                    .toUriString();
            val response = restTemplate.getForObject(url, OpenMeteoAirQuality.class);
            log.debug("Response from Open-Meteo Air Quality API: {}", response);
            if (response == null || response.getCurrent() == null) {
                return null;
            }
            return weatherMapper.toAirQualityData(response);
        } catch (Exception e) {
            log.error("Error fetching air quality data for lat: {}, lon: {}: {}", latitude, longitude, e.getMessage());
            return null;
        }
    }
}

File: src/main/webapp/src/modules/weather/service/weatherService.ts
import {
    type AirQualityData,
    Configuration,
    type GeocodingResult,
    type Location,
    WeatherApi,
    type WeatherForecast,
    WeatherProvider,
} from '@api';
import i18n from '../../../i18n/i18n.ts';
import type { WeatherSettings } from '../types/types.ts';

// Create API instances
const configuration = new Configuration({
    basePath: '/daily/api',
});

const weatherApi = new WeatherApi(configuration);

export class WeatherService {
    private readonly city: string;
    private latitude: number | null = null;
    private longitude: number | null = null;
    private location: Location | null = null;
    private initialized = false;
    private provider: WeatherProvider | null = null;
    private locationKey: string | null = null;

    constructor(settings: WeatherSettings) {
        this.city = settings.city || '';
        this.latitude = settings.latitude || null;
        this.longitude = settings.longitude || null;
        this.provider = settings.provider || null;
        this.locationKey = settings.locationKey || null;
    }

    // Initialize service and get location data
    public async initialize(): Promise<void> {
        if (this.initialized) {
            return;
        }

        try {
            // Require city and provider to be present
            if (!this.city) {
                throw new Error('City name is required for initialization');
            }
            if (!this.provider) {
                throw new Error('Weather provider must be selected for initialization');
            }

            // Always resolve coordinates (and possibly locationKey) from city + provider
            const geocodingResult = await this.getGeocodingData(this.city, this.provider);
            if (!geocodingResult) {
                throw new Error(`Geocoding data not found for city: ${this.city}`);
            }
            this.latitude = geocodingResult.latitude || 0;
            this.longitude = geocodingResult.longitude || 0;
            // persist locationKey if returned (important for AccuWeather)
            this.locationKey = geocodingResult.locationKey || this.locationKey;

            // Create location data from geocoding result
            this.location = {
                id: geocodingResult.name,
                name: geocodingResult.name || '',
                country: geocodingResult.country || '',
                state: geocodingResult.state || '',
                lat: geocodingResult.latitude,
                lon: geocodingResult.longitude,
                locationKey: geocodingResult.locationKey,
            };

            this.initialized = true;
        } catch (error) {
            console.error('Error initializing weather service:', error);
            throw error;
        }
    }

    // Get cached location data or reinitialize if needed
    public async getLocationData(): Promise<Location> {
        if (!this.initialized || !this.location) {
            await this.initialize();
        }
        return this.location as Location;
    }

    // Get geocoding data for a city name
    private async getGeocodingData(
        cityName: string,
        provider?: WeatherProvider,
    ): Promise<GeocodingResult | null> {
        if (!cityName) {
            console.warn('City name is empty, cannot fetch geocoding data');
            return Promise.resolve(null);
        }
        try {
            // Use weatherApi to geocode city name with correct parameter syntax
            const response = await weatherApi.geocodeLocation(cityName, i18n.language, provider);
            return response.data || null;
        } catch (error) {
            console.error('Error fetching geocoding data:', error);
            return null;
        }
    }

    public async getWeatherForecast(): Promise<WeatherForecast | null> {
        if (!this.provider) {
            console.warn('Weather provider is not set, cannot fetch forecast');
            return null;
        }
        switch (this.provider) {
            case WeatherProvider.Openweather:
                return this.getOpenWeatherForecast();
            case WeatherProvider.Accuweather:
                return this.getAccuWeatherForecast();
            default:
                console.warn('Unsupported weather provider:', this.provider);
                return null;
        }
    }

    // Get current, hourly, and daily weather data (OpenWeather)
    public async getOpenWeatherForecast(): Promise<WeatherForecast | null> {
        // Ensure service is initialized
        if (!this.initialized) {
            await this.initialize();
        }

        // Provider must be OpenWeather for this method
        if (this.provider !== WeatherProvider.Openweather) {
            console.warn('OpenWeather forecast requested but provider is not OpenWeather');
            return null;
        }

        // Latitude and longitude are required
        if (!this.latitude || !this.longitude) {
            console.warn('Latitude and longitude are required to fetch OpenWeather forecast');
            return null;
        }

        try {
            // Call the API to get forecast data
            const response = await weatherApi.getOpenWeatherForecast(
                this.latitude,
                this.longitude,
                5,
            );
            return response.data || null;
        } catch (error) {
            console.error('Error fetching OpenWeather forecast data:', error);
            return null;
        }
    }

    // Get AccuWeather forecast using locationKey
    public async getAccuWeatherForecast(): Promise<WeatherForecast | null> {
        // Ensure service is initialized
        if (!this.initialized) {
            await this.initialize();
        }

        // Provider must be AccuWeather for this method
        if (this.provider !== WeatherProvider.Accuweather) {
            console.warn('AccuWeather forecast requested but provider is not AccuWeather');
            return null;
        }

        // locationKey is required for AccuWeather
        const keyFromLocation = this.location?.locationKey || this.locationKey;
        if (!keyFromLocation) {
            console.warn(
                'AccuWeather provider selected but no locationKey is available; ensure geocoding provided a locationKey',
            );
            return null;
        }

        const numericKey = Number(keyFromLocation);
        if (!Number.isFinite(numericKey)) {
            console.error('Invalid AccuWeather locationKey, expected a numeric value');
            return null;
        }

        try {
            const response = await weatherApi.getAccuWeatherForecast(numericKey, 5);
            return response.data || null;
        } catch (error) {
            console.error('Error fetching AccuWeather forecast data:', error);
            return null;
        }
    }

    // Get air quality data
    public async getAirQuality(): Promise<AirQualityData | null> {
        if (!this.latitude || !this.longitude) {
            await this.initialize();
            if (!this.latitude || !this.longitude) {
                return null;
            }
        }
        try {
            // Call the API to get air quality data
            const response = await weatherApi.getAirQuality(this.latitude, this.longitude);
            return response.data || null;
        } catch (error) {
            console.error('Error fetching air quality data:', error);
            return null;
        }
    }
}
